sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1843,11 +1844,37 @@ public PCollection<KafkaRecord<K, V>> expand(PBegin input) {
                       kafkaRead.getStartReadTime(),
                       kafkaRead.getStopReadTime()));
         } else {
-          output =
+          // Use Impulse for backwards compatibility with previous released versions.
+          String requestedVersionString =
               input
                   .getPipeline()
-                  .apply(Impulse.create())
-                  .apply(ParDo.of(new GenerateKafkaSourceDescriptor(kafkaRead)));
+                  .getOptions()
+                  .as(StreamingOptions.class)
+                  .getUpdateCompatibilityVersion();
+          boolean useImpulse = false;
+          if (requestedVersionString != null) {
+            List<String> requestedVersion = Arrays.asList(requestedVersionString.split("\\."));
+            List<String> targetVersion = Arrays.asList("2", "66", "0");
+
+            if (Comparators.lexicographical(Comparator.<String>naturalOrder())
+                    .compare(requestedVersion, targetVersion)
+                < 0) {
```

Comment:
There's a helper function called `TransformUpgrader.compareVersions` that's used in `KafkaIOTranslation`. I know the other use of update compatibility version comparison in this file uses the same approach so it's fine by me if it stays like this.

Reply:
Thanks! I looked a little but didn't find that, I think it's clearer to use the one-liner so I updated both sites in this PR.